src/Sylius/Bundle/AddressingBundle/Validator/Constraints/ProvinceAddressConstraintValidator.php

```diff
@@ -47,7 +47,7 @@ public function validate($value, Constraint $constraint): void
         $propertyPath = $this->context->getPropertyPath();
 
         foreach (iterator_to_array($this->context->getViolations()) as $violation) {
-            if (0 === strpos($violation->getPropertyPath(), $propertyPath)) {
+            if ('' === $propertyPath || 0 === strpos($violation->getPropertyPath(), $propertyPath)) {
```

Comment:
Wouldn't that skip the validation if you pass to the validator an instance of Address class and there's a violation before this constraint is validated?
The property path is an empty string when the root value is being validated.
The way you wrote the condition (`'' === $propertyPath ||`), will check if the property path is not an empty string first, which will always be true when the root object is being validated, so it won't even evaluate the left side of the `||`.
If there's is a constraint before this one that results in a violation, then it will enter the loop and will return immediately.

I suggest changing the condition as show below.
Also, a test is definitely necessary here, especially for the unfortunate scenario I described.
Can you, please, change the condition and also update the PHP Spec file, @rimas-kudelis?

```suggestion
            if ('' !== $propertyPath && 0 === strpos($violation->getPropertyPath(), $propertyPath)) {
```

Reply:
> Wouldn't that skip the validation if you pass to the validator an instance of Address class and there's a violation before this constraint is validated?

Like I mentioned in the initial comment, I suppose that's the whole point of this condition in the first place: if an address is deemed invalid already, this validator does not perform further checks. The only problem was the root object case, due to how `strpos()` works.

Reply:
> Also, a test is definitely necessary here

Good point! I was reluctant to add one, because I don't have a quick way to run them. But t I've added one now (which iis mostly a copy of one that already exists).

Reply:
@vvasiloi I'll leave it for you to mark this conversation as resolved if you agree with me.

Reply:
Thank you for the clarification!
Unfortunately, now I have more questions about the whole `foreach` block... 😅 
I wish there would be some unit tests for it, because the spec file doesn't answer those questions.
Speaking of spec, I have a feeling that the test you added will give false-positive results, just like the one you copied it from.
It expects that `addViolation` should not be called, but without proper setup on the address, it will never be called, because the call is wrapped in an `if` statement that does the actual validation.
You can probably remove the foreach and the test will still pass...
WDYT?

Reply:
@vvasiloi the way I understand it is:
- This validator checks an Address **Object**, because it needs to look at two of its fields: `country` and `province`
- Before doing the actual validation, it loops (hence the `foreach`) through the violations already in the context, and checks if there's at least one added to either **the Address object itself**, or **one of its properties** (hence the `strpos` instead of a simple comparison). If it finds at least one violation on the address/its property, it skips its own validation.

For example, if you're validating an order that is being submitted, and a shipping addess has an invalid country, or province (or actually any property that is invalid), that violation will have a property path like `shipping_address.country`, and this validator, which will be defined with a `shipping_address` property path will return early.

Similarly, if the address itself already has a violation (property path `shipping_address`), that will also match the `strpos()` test and cause an early return.

On the other hand, if all violations will have unrelated property paths (e.g. `billing_address`), this validator **will** perform its check, because it's property path `shipping_address` will not match the beginning of all these property paths with which the violations are added.

But the problem arises when our property path is empty, because `strpos('some_path', '')` does not simply return `0` or `false`, but triggers an error due to its second argument (`needle`) being an empty string (because, it's indeed weird to test whether an empty string is a substring of any other string).  That's what my patch fixes.

You're probably right about the test though: it seems rather superfluous and would probably not fail even without the `foreach` loop. Hmm...

Reply:
It's probably fine, no need to complicate matters.
I suggested a change for the test, take a look.